File: s1-web/client/src/VM.js
import { loxValueToString, shortLittleEndianToInteger } from "./utils";

const TOP_LEVEL_SCRIPT = {
  functionName: "__toplevel__",
};

function falsey(loxValue) {
  return !loxValue;
}

function equal(firstLoxValue, secondLoxValue) {
  return firstLoxValue === secondLoxValue;
}

function createCallable(functionDescriptor) {
  return {
    functionName: functionDescriptor.name,
    arity: functionDescriptor.arity,
  };
}

function createCallFrame(executable, stack, callable, stackTop) {
  let ip = 0;

  return {
    stackTop,
    functionName: callable.functionName,
    callable,
    readCode() {
      ip += 1;
      return executable.functions[callable.functionName][ip - 1];
    },
    ip() {
      return ip;
    },
    readShort() {
      return shortLittleEndianToInteger(this.readCode(), this.readCode());
    },

    peekCode() {
      return executable.functions[callable.functionName][ip];
    },

    readConstant(constantIndex) {
      return executable.constants[constantIndex];
    },

    getStackSlot(offset) {
      return stack[stackTop + offset];
    },

    setStackSlot(offset, value) {
      stack[stackTop + offset] = value;
    },

    jump(offset) {
      ip += offset;
    },
  };
}

export class VM {
  constructor(executable) {
    this.executable = executable;
    this.reset();
  }

  reset() {
    this.output = [];
    this.stack = [];
    this.globals = {};
    this.callFrames = [
      createCallFrame(this.executable, this.stack, TOP_LEVEL_SCRIPT, {}, 0),
    ];
  }

  step() {
    const callFrame = this.callFrames[this.callFrames.length - 1];

    const op = callFrame?.readCode();

    if (callFrame && op) {
      switch (op) {
        case "DEFINE-GLOBAL":
          this.globals[callFrame.readConstant(callFrame.readCode())] =
            this.stack.pop();
          break;
        case "GET-GLOBAL":
          this.stack.push(
            this.globals[callFrame.readConstant(callFrame.readCode())]
          );
          break;
        case "SET-GLOBAL":
          this.globals[callFrame.readConstant(callFrame.readCode())] =
            this.stack.pop();
          break;
        case "GET-LOCAL":
          this.stack.push(callFrame.getStackSlot(callFrame.readCode()));
          break;
        case "SET-LOCAL":
          callFrame.setStackSlot(
            callFrame.readCode(),
            this.stack[this.stack.length - 1]
          );
          break;
        case "EQUAL": {
          const b = this.stack.pop();
          const a = this.stack.pop();
          this.stack.push(equal(a, b));
          break;
        }
        case "LESSER": {
          const b = this.stack.pop();
          const a = this.stack.pop();
          this.stack.push(a < b);
          break;
        }
        case "GREATER": {
          const b = this.stack.pop();
          const a = this.stack.pop();
          this.stack.push(a > b);
          break;
        }
        case "NOT":
          this.stack.push(!this.stack.pop());
          break;
        case "LOAD-CONSTANT":
          this.stack.push(callFrame.readConstant(callFrame.readCode()));
          break;
        case "LOAD-CLOSURE":
          const functionDescriptor = callFrame.readConstant(
            callFrame.readCode()
          );

          const callable = createCallable(functionDescriptor);
          this.stack.push(callable);
          break;
        case "ADD": {
          const b = this.stack.pop();
          const a = this.stack.pop();
          this.stack.push(a + b);
          break;
        }
        case "SUBTRACT": {
          const b = this.stack.pop();
          const a = this.stack.pop();
          this.stack.push(a - b);
          break;
        }
        case "MULTIPLY": {
          const b = this.stack.pop();
          const a = this.stack.pop();
          this.stack.push(a * b);
          break;
        }
        case "DIVIDE": {
          const b = this.stack.pop();
          const a = this.stack.pop();
          this.stack.push(a / b);
          break;
        }
        case "PRINT":
          this.output.push(loxValueToString(this.stack.pop()));
          break;
        case "POP":
          this.stack.pop();
          break;
        case "NIL":
          this.stack.push(null);
          break;
        case "TRUE":
          this.stack.push(true);
          break;
        case "NEGATE":
          this.stack.push(-this.stack.pop());
          break;
        case "FALSE":
          this.stack.push(false);
          break;
        case "JUMP-ON-FALSE":
          const offset = callFrame.readShort();
          if (falsey(this.stack[this.stack.length - 1])) callFrame.jump(offset);
          break;
        case "JUMP":
          callFrame.jump(callFrame.readShort());
          break;
        case "CALL": {
          const argumentCount = callFrame.readCode();
          const callable = this.stack[this.stack.length - argumentCount - 1];

          const newCallFrame = createCallFrame(
            this.executable,
            this.stack,
            callable,
            this.stack.length - argumentCount
          );

          this.callFrames.push(newCallFrame);
          break;
        }
        case "RETURN":
          const result = this.stack.pop();
          this.callFrames.pop();

          while (
            this.stack.length > 0 &&
            this.stack.length >= callFrame.stackTop
          )
            this.stack.pop();

          if (this.callFrames.length > 0) {
            this.stack.push(result);
          }
          break;
        default:
          break;
      }
    }
  }

  currentState() {
    return {
      output: this.output,
      stack: this.stack,
      globals: this.globals,
      callFrames: this.callFrames,
      callFrame: this.callFrames[this.callFrames.length - 1],
      nextOp: this.callFrames[this.callFrames.length - 1]?.peekCode(),
      terminated: this.callFrames.length === 0,
    };
  }

  run() {
    while (this.callFrames.length > 0) {
      this.step();
    }
  }
}

File: s1-web/a_lox/lib/a_lox/vm.rb
require "ostruct"

module ALox
  class VM
    class Function
      def self.top_level_script
        OpenStruct.new(
          function_name: "__toplevel__",
          upvalues: []
        )
      end

      attr_reader :upvalues

      def initialize(compiled_function, upvalues)
        @compiled_function = compiled_function
        @upvalues = upvalues
      end

      def function_name
        @compiled_function.name
      end

      def arity
        @compiled_function.arity
      end
    end

    class CallFrame
      attr_reader :function, :stack_top

      def initialize(executable, stack, function, stack_top = 0)
        @executable = executable
        @stack = stack
        @function = function
        @ip = 0
        @stack_top = stack_top
      end

      def read_code
        @ip += 1
        @executable.functions[@function.function_name][@ip - 1]
      end

      def read_short
        BinaryUtils.unpack_short(read_code, read_code)
      end

      def read_constant(constant_index)
        @executable.constants[constant_index]
      end

      def get_stack_slot(offset)
        @stack[@stack_top + offset]
      end

      def set_stack_slot(offset, value)
        @stack[@stack_top + offset] = value
      end

      def jump(location)
        @ip += location
      end
    end

    class Upvalue
      attr_reader :pointer

      def initialize(pointer, stack)
        @pointer = pointer
        @stack = stack
      end

      def close!
        @value = @stack[pointer]
      end

      def value
        @value || @stack[pointer]
      end

      def set_value(new_value)
        if defined? @value
          @value = new_value
        else
          @stack[pointer] = new_value
        end
      end
    end

    def self.execute(executable, out: $stdout, debug: false)
      new.execute(executable, out: out, debug: debug)
    end

    def initialize(error_reporter: nil)
      @stack = []
      @globals = {}
      @error_reporter = error_reporter
    end

    def execute(executable, out: $stdout, debug: false)
      call_frames = [
        CallFrame.new(executable, @stack, Function.top_level_script)
      ]

      open_upvalues = []

      @had_error = false


      loop do
        call_frame = call_frames.last

        break if call_frame.nil?
        break if @had_error

        op = call_frame.read_code

        case op
        when Opcodes::LOAD_CONSTANT
          @stack.push(call_frame.read_constant(call_frame.read_code))
        when Opcodes::LOAD_CLOSURE
          compiled_function = call_frame.read_constant(call_frame.read_code)

          upvalues =
            compiled_function.upvalue_count.times.map do |upvalue_description|
              is_local, slot = call_frame.read_code, call_frame.read_code

              if is_local == 1
                new_upvalue =
                  open_upvalues.detect { _1.pointer == call_frame.stack_top + slot } ||
                    Upvalue.new(call_frame.stack_top + slot, @stack)

                open_upvalues << new_upvalue
                open_upvalues.sort_by! { -(_1.pointer) }

                new_upvalue
              else
                call_frame.function.upvalues[slot]
              end
            end

          @stack.push(
            Function.new(compiled_function, upvalues)
          )
        when Opcodes::SET_GLOBAL
          @globals[call_frame.read_constant(call_frame.read_code)] = @stack.last
        when Opcodes::DEFINE_GLOBAL
          @globals[call_frame.read_constant(call_frame.read_code)] = @stack.pop
        when Opcodes::GET_GLOBAL
          global_name = call_frame.read_constant(call_frame.read_code)
          error("Undefined global #{global_name}") unless @globals.has_key?(global_name)
          @stack.push(@globals[global_name])
        when Opcodes::SET_LOCAL
          call_frame.set_stack_slot(call_frame.read_code, @stack.last)
        when Opcodes::GET_LOCAL
          @stack.push(call_frame.get_stack_slot(call_frame.read_code))
        when Opcodes::GET_UPVALUE
          @stack.push(call_frame.function.upvalues[call_frame.read_code].value)
        when Opcodes::SET_UPVALUE
          call_frame.function.upvalues[call_frame.read_code].set_value(@stack.last)
        when Opcodes::NIL_OP
          @stack.push(nil)
        when Opcodes::TRUE_OP
          @stack.push(true)
        when Opcodes::FALSE_OP
          @stack.push(false)
        when Opcodes::NOT
          @stack.push(falsey?(@stack.pop))
        when Opcodes::NEGATE
          @stack.push(-@stack.pop)
        when Opcodes::POP
          fail 'corrupted stack' if @stack.empty?
          @stack.pop
        when Opcodes::ADD
          b, a = @stack.pop, @stack.pop
          @stack.push(a + b)
        when Opcodes::SUBTRACT
          b, a = @stack.pop, @stack.pop
          @stack.push(a - b)
        when Opcodes::DIVIDE
          b, a = @stack.pop, @stack.pop
          @stack.push(a / b)
        when Opcodes::MULTIPLY
          b, a = @stack.pop, @stack.pop
          @stack.push(a * b)
        when Opcodes::EQUAL
          @stack.push(equal?(@stack.pop, @stack.pop))
        when Opcodes::GREATER
          b, a = @stack.pop, @stack.pop
          @stack.push(a > b)
        when Opcodes::LESSER
          b, a = @stack.pop, @stack.pop
          @stack.push(a < b)
        when Opcodes::CALL
          argument_count = call_frame.read_code
          function = @stack[-argument_count - 1]

          if function.is_a? Function
            if function.arity == argument_count
              call_frames << CallFrame.new(
                executable,
                @stack,
                function,
                @stack.size - argument_count
              )
            else
              error("function #{function.function_name} takes #{function.arity} arguments but #{argument_count} provided")
            end
          else
            error("#{lox_object_to_string(function)} is not function")
          end
        when Opcodes::RETURN
          result = @stack.pop

          while open_upvalues.count > 0 && open_upvalues.first.pointer >= call_frame.stack_top
            upvalue = open_upvalues.shift
            upvalue.close!
          end

          call_frames.pop

          @stack[0..-1] = @stack[0..call_frame.stack_top - 2]

          @stack.push(result)
        when Opcodes::CLOSE_UPVALUE
          fail 'corrupted upvalues' if open_upvalues.empty?
          open_upvalues.shift.close!

          fail 'corrupted stack' if @stack.empty?
          @stack.pop
        when Opcodes::PRINT
          out.puts(lox_object_to_string(@stack.pop))
        when Opcodes::JUMP_ON_FALSE
          location = call_frame.read_short
          call_frame.jump(location) if falsey?(@stack.last)
        when Opcodes::JUMP
          call_frame.jump(call_frame.read_short)
        else
          fail op.inspect
        end

        print_debug_info(op) if debug
      end

      @stack.pop

      fail 'corrupted open upvalues' if open_upvalues.any?
    end

    private

    def falsey?(value)
      !value
    end

    def equal?(a, b)
      a == b # TODO
    end

    def lox_object_to_string(lox_object)
      if lox_object.is_a? Function
        "fun #{lox_object.function_name}/#{lox_object.arity}"
      elsif lox_object.nil?
        "nil"
      else
        lox_object.to_s
      end
    end

    def error(message)
      @had_error = true
      @error_reporter&.report_runtime_error(message)
    end

    def print_debug_info(op)
      puts "DEBUG: #{op}"
      print_stack
      puts
    end

    def print_stack
      puts [
        "STACK:",
        "[",
        *@stack.map { |value| value.is_a?(Function) ? "<#{value.function_name}>" : value.inspect }.join(", "),
        "]"
      ].join(" ")
    end
  end
end
